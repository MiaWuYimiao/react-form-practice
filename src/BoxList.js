import React, { useState } from "react";
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from "uuid";

const BoxList = () => {
    const initialBoxes = [{
        id: uuid(),
        backgroundColor: "blue",
        width: 100,
        height: 50
    }, {
        id: uuid(),
        backgroundColor: "red",
        width: 20,
        height: 60
    }]

    const [boxes, setBoxes] = useState([]);

    const addBox = (box) => {
        let newBox = { ...box, id:uuid() };
        setBoxes(oldBoxes => [...oldBoxes, newBox]);
    };

    const deleteBox = (id) => {
        setBoxes(oldBoxes => (
            oldBoxes.filter(box => box.id != id)
        ))
    };

    return (
        <div>
            {boxes.map(box => (
                <Box backgroundColor={box.backgroundColor}
                    id={box.id}
                    height={box.height}
                    width={box.width}
                    key={box.id}
                    deleteBox={deleteBox}
                />
            ))}
            {<NewBoxForm addBox={addBox}/>}
        </div>
    )
}

export default BoxList;