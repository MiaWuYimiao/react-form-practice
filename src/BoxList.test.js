import React from "react";
import {fireEvent, render, wait} from "@testing-library/react"
import BoxList from "./BoxList";

it("render without crash", function() {
    render(<BoxList  />);
})

it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

function addBox(boxList, height=2, width=2, color="peachpuff") {
    const heightInput = boxList.getByLabelText("Height:");
    const widthInput = boxList.getByLabelText("Width:");
    const backgroundInput = boxList.getByLabelText("BackgroundColor:");
    fireEvent.change(heightInput, { target: {value:height}});
    fireEvent.change(widthInput, { target: {value:width}});
    fireEvent.change(backgroundInput, { target: {value:color}});
    const button = boxList.getByText("Add");
    fireEvent.click(button);
}

it("can add a new box", function() {
    const boxList = render(<BoxList />);

    expect(boxList.queryByText("X")).not.toBeInTheDocument();

    addBox(boxList);

    const removeButton = boxList.getByText("X");
    expect(removeButton).toBeInTheDocument();
    expect(removeButton.previousSibling).toHaveStyle(`
        width: 2px;
        height: 2px;
        background-color: peachpuff;
    `)
    expect(boxList.getAllByDisplayValue("")).toHaveLength(3);
});

it("can remove a box", function() {
    const boxList = render(<BoxList />);
    addBox(boxList);

    const removeButton = boxList.getByText("X");

    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
})