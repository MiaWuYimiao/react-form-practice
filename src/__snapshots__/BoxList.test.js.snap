// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <div>
          <label
            for="backgroundColor"
          >
            BackgroundColor:
          </label>
          <input
            id="backgroundColor"
            name="backgroundColor"
            placeholder="backgroundColor"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            for="width"
          >
            Width:
          </label>
          <input
            id="width"
            name="width"
            placeholder="width"
            type="Number"
            value=""
          />
        </div>
        <div>
          <label
            for="height"
          >
            Height:
          </label>
          <input
            id="height"
            name="height"
            placeholder="height"
            type="Number"
            value=""
          />
        </div>
        <button>
          Add
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
